vllm/worker/model_runner.py

```diff
@@ -646,82 +655,94 @@ def _prepare_model_input(
             num_prefills=num_prefills,
         )
 
-    def prepare_input_tensors(
-        self,
-        seq_group_metadata_list: Optional[List[SequenceGroupMetadata]],
-    ) -> Tuple[torch.Tensor, torch.Tensor, AttentionMetadata, SamplingMetadata,
-               Set[LoRARequest], LoRAMapping, Dict[str, torch.Tensor]]:
-        if self.is_driver_worker:
-            assert seq_group_metadata_list is not None
-            # Prepare input tensors.
-            (
-                input_tokens,
-                input_positions,
-                attn_metadata,
-                seq_lens,
-                query_lens,
-                lora_mapping,
-                lora_requests,
-                multi_modal_kwargs,
-                slot_mapping,
-                num_prefill_tokens,
-                num_decode_tokens,
-                num_prefills,
-            ) = self._prepare_model_input(seq_group_metadata_list)
-            sampling_metadata = SamplingMetadata.prepare(
-                seq_group_metadata_list, seq_lens, query_lens, self.device,
-                self.pin_memory)
-
-            metadata_dict = {
-                "input_tokens": input_tokens,
-                "input_positions": input_positions,
-                "selected_token_indices":
-                sampling_metadata.selected_token_indices,
-                "lora_requests": lora_requests,
-                "lora_mapping": lora_mapping,
-                "multi_modal_kwargs": multi_modal_kwargs,
-                "num_prefill_tokens": num_prefill_tokens,
-                "num_decode_tokens": num_decode_tokens,
-                "slot_mapping": slot_mapping,
-                "num_prefills": num_prefills,
-            }
-            if attn_metadata:
-                metadata_dict.update(attn_metadata.asdict_zerocopy())
-            broadcast_tensor_dict(metadata_dict, src=0)
+    def prepare_inputs_to_broadcast(
```

Comment:
I wonder whether we should have this return `ModelRunnerInput` and then add a method to that `get_dict_to_broadcast()` which you only call in the TP > 1 case. This would be more efficient for the non-TP case.

Then this method could be called `prepare_modelrunner_input()` which would make more sense to me since it's used even for non-TP where broadcasting isn't being done.

That would also obviate the need for this separate `aux` arg/variable.

Reply:
The problem is who takes the responsibility to drive the broadcast process and how does it know the function to call, given that we have inheritance in `model_runner.py` and `embedding_model_runner.py` .

Previously, each model runner drives the broadcast operation itself, inside `execute_model`, which leads to a separate broadcast.

In this PR, worker drives the broadcast operation, so it needs `ModelRunner` to return `inputs_to_broadcast`, broadcast it, and then feed it back to `ModelRunner`.

If we use `ModelRunnerInput. get_dict_to_broadcast `, how can the worker know which `ModelRunnerInput` to use? We have multiple `ModelRunnerInput`, used in both `model_runner.py` and `embedding_model_runner.py`.

Reply:
> If we use `ModelRunnerInput. get_dict_to_broadcast `, how can the worker know which `ModelRunnerInput` to use? We have multiple `ModelRunnerInput`, used in both `model_runner.py` and `embedding_model_runner.py`.

You kind of answered your own question :) ... both `ModelRunnerInput`s implement that and the worker just calls it (could be a [protocol](https://peps.python.org/pep-0544/))

Reply:
this can be solved by adding an abstract class and let ModelRunnerInput inherit the abstract class. For reasons I don't know, inheritance is discouraged in vllm, and several modelrunners just duplicate the code.

Reply:
Inheritance is used in various places already. IMO it makes sense to use judiciously but to avoid overdoing it.

In any case that's why I suggested to use a [protocol](https://peps.python.org/pep-0544/) here, you can do the same thing without inheritance.